2024Robot_Team3484/src/main/cpp/subsystems/LauncherSubsystem.cpp

```diff
@@ -40,36 +40,56 @@ LauncherSubsystem::LauncherSubsystem(
-        Launcher_m_Left_pidController->SetI(I_Launcher);
-        Launcher_m_Left_pidController->SetD(D_Launcher);
+        Launcher_m_Left_pidController->SetP(pidc.Kp);
+        Launcher_m_Left_pidController->SetI(pidc.Ki);
+        Launcher_m_Left_pidController->SetD(pidc.Kd);
         Launcher_m_Left_pidController->SetIZone(0);
-        Launcher_m_Left_pidController->SetFF(FF_Launcher);
+        Launcher_m_Left_pidController->SetFF(pidc.Kf);
         Launcher_m_Left_pidController->SetOutputRange(0, 1);
         }
     if (Launcher_m_Right_pidController !=NULL){
-        Launcher_m_Right_pidController->SetP(P_Launcher);
-        Launcher_m_Right_pidController->SetD(D_Launcher);
-        Launcher_m_Right_pidController->SetI(I_Launcher);
+        Launcher_m_Right_pidController->SetP(pidc.Kp);
+        Launcher_m_Right_pidController->SetI(pidc.Ki);
+        Launcher_m_Right_pidController->SetD(pidc.Kd);
         Launcher_m_Right_pidController->SetIZone(0);
-        Launcher_m_Right_pidController->SetFF(FF_Launcher);
+        Launcher_m_Right_pidController->SetFF(pidc.Kf);
         Launcher_m_Right_pidController->SetOutputRange(0, 1);
     }
 }
 
 void LauncherSubsystem::setLauncherRPM(units::revolutions_per_minute_t speed){
-    _target_speed = speed.value();
+    _target_speed = speed.value()*GEAR_RATIO;
 }
 
 void LauncherSubsystem::Periodic() {
+    #ifdef EN_DIAGNOSTICS
+        SmartDashboard::PutNumber("Motor Speed Left (RPM)", Launcher_Encoder_Left->GetVelocity()/GEAR_RATIO);
+        SmartDashboard::PutNumber("Motor Speed Right (RPM)", Launcher_Encoder_Right->GetVelocity()/GEAR_RATIO);
+    #endif
+    _counter_null_right = 0;
+    _counter_null_left = 0;
+
     if (Launcher_m_Left_pidController !=NULL){
         Launcher_m_Left_pidController->SetReference(_target_speed, rev::CANSparkMax::ControlType::kVelocity);
+        _counter_null_left++;
     }
      if (Launcher_m_Right_pidController !=NULL){
         Launcher_m_Right_pidController->SetReference(_target_speed, rev::CANSparkMax::ControlType::kVelocity);
+        _counter_null_right++;
      }
 }
 
 bool LauncherSubsystem::atTargetRPM(){
- return std::abs(Launcher_Encoder_Left->GetVelocity()-_target_speed) < RPM_Window_Launcher && std::abs(Launcher_Encoder_Right->GetVelocity()-_target_speed) < RPM_Window_Launcher;   
+    if (_counter_null_left + _counter_null_right == 0){
+        return std::abs(Launcher_Encoder_Left->GetVelocity()-_target_speed) < RPM_WINDOW_RANGE && std::abs(Launcher_Encoder_Right->GetVelocity()-_target_speed) < RPM_WINDOW_RANGE;   
+    }
+    else if (_counter_null_left == 1) {
+        return std::abs(Launcher_Encoder_Left->GetVelocity()-_target_speed) < RPM_WINDOW_RANGE;   
+    }
+    else if (_counter_null_right == 1) {
+        return std::abs(Launcher_Encoder_Right->GetVelocity()-_target_speed) < RPM_WINDOW_RANGE;   
+    }
+    if (_counter_null_left + _counter_null_right == 2){
+        return false;
```

Comment:
Based on how your null counter logic works in the periodic function. You will be returning `FALSE` when both PID controllers are NOT null, and the `atTargetRPM` function will only return `TRUE` when at least 1 of the PID controllers is null. To fix this you can either:
a) Change the first `if` condition to check if your null counters are equal to 2.
b) Move the counter increment statements in `Periodic()` to an `else` condition of the PID controller null checks.

```
if (both are not null)
{
    return true if both PID controllers are in range
}
else if (right is null)
{
    return true if left PID controller is in range
}
else if (left is null)
{
    return true if right PID controller is in range
}
else
{
    return false;
}
```

Also, I suggest having the `return false;` statement be an `else` condition in the `atTargetRPM` function.